### PR TITLE
Billing Manager Update

### PR DESCRIPTION
**Pull request by sjpadgett**
- redo drop down lists button styles to links
- added new dry run validate claims action
- fix clear to not update claim after downloads after setting billed flag and clearing processing dates
- fix cancel to cancel action
- fix payer_type
- update jquery to 3.1.1 version

below added by bradymiller:
Up For Grabs for this PR is here: https://www.open-emr.org/wiki/index.php/Development_Demo#Zeta_-_Up_For_Grabs_Demo

### CODE REVIEW

**Comment by sjpadgett** [comment]
Will leave open for awhile for folks to beat up..

File: interface/billing/billing_report.php [hunk 1]
```diff
@@ -109,14 +109,137 @@
...
 </style>
 <script>
     var partners = <?php echo json_encode($partners); ?>;
+    // next set of 4 functions are for a wait confirm action then submit.
+    // I wrote this a little more involved than it needs to be
+    // to example the pattern. Ideal submit part would be via a fetch or ajax
+    // then could do refresh or after submit actions.
+    function doSubmit() {
+        top.restoreSession();
+        return new Promise(function(resolve, reject){
```
Comment (bradymiller) [inline interface/billing/billing_report.php]: woohooo, another Promise ! :)

**Comment by bradymiller** [comment]
I sadly was unable to find any escaping issues with this code. :)
Code looks good to me and I'll get a demo going for folks to be able to test, if needed.

**Comment by sjpadgett** [comment]
The neat part is really using async/await. can be very handy.
Demo would be nice so Sherwin and Stephen can ponder...

**Comment by bradymiller** [comment]
Up For Grabs for this PR is here: https://www.open-emr.org/wiki/index.php/Development_Demo#Zeta_-_Up_For_Grabs_Demo

**Comment by stephenwaite** [comment]
will check it out with my billing database this afternoon, thank you @sjpadgett

**Comment by stephenwaite** [comment]
not seeing where claims status is set to billed after clicking Validate and Clear - "Validate and Clear validates then resets claims status to billed and clears process time"

**Comment by sjpadgett** [comment]
That's because page doesn't refresh. Never has I know of if a download is selected.

**Comment by sjpadgett** [comment]
It works for me if I click update and billing table is updated to billed. What am I missing @stephenwaite

**Comment by stephenwaite** [comment]
common use is to go billing manager and remove date of service today and click update list to see what is unbilled

if do that and then bill a claim shouldn't "Validate and Clear" change it to billed so when come back to billing manager and click update list it is now gone?

**Comment by sjpadgett** [comment]
Yes but what i'm saying is it works for me. i've tested three times.

**Comment by stephenwaite** [comment]
ok, darn, sorry, will rebuild dev environment retest

**Comment by stephenwaite** [comment]
not working here https://four.openemr.io/c/openemr/interface/login/login.php?site=default

**Comment by stephenwaite** [comment]
maybe it's this message "Validate and Clear validates then resets claims status to billed and clears process time" is it resetting the claim status to what it was before running Validate and Clear?

**Comment by sjpadgett** [comment]
Clear will set claim status to 2 (billed), billed flag to 1 (in billing table), payer_type to current in UI (primary etc), set bill_process to 0 and clear process date and target.
Otherwise, continue/normal process will set claim status to 2 (billed), billed flag to 1, payer_type to -1, bill_process to 2.
You're right it's not working on linux

**Comment by sjpadgett** [comment]
Turns out that it works as before however, while claim status is set to billed, claims bill_process is set to 0 and billings billed flag is set to 1, billing table bill_process is left unaltered(generally 2).  The billings bill_process is what determines if claim is billed(we should really call it submitted) ie a claim could be marked as billed but not submitted which is what the clear flag is doing however, while the claim bill_process status is changed, billing is not so there must be a purpose in posting for this difference else why bother.

I don't know what to do about this clear stuff. I just can't see a purpose for it!!!!

**Comment by zbig01** [comment]
Technically speaking the `X12 Options` and` HCFA Form` are Bootstrap dropdown buttons and not links per-se. They lead to options that are also buttons which are again not links but result in downloading the appropriate file. It would be semantically correct to leave these two dropdown buttons with `btn-default` class. Styling the `<li>` buttons for the dropdown buttons with `btn-link` class would be more appropriate if you desire to give it a more dropdown menu like appearance. The downloadable nature of these buttons is emphasized by the download icon on each button. By leaving the two dropdown buttons with `btn-default` class you give equal weight to all four top-level buttons -` X12 Options, HCFA Form, Mark As Cleared` and `Re-Open`. All four buttons start off disabled and when the check box is checked the all relevant buttons become enabled. Using `btn-deafult` on all four top-level buttons ensures that all the enabled buttons will have similar emphasis.

The difference is more noticeable if a color theme is used.

**Light Theme - class btn-link - options disabled**

billing_manager_01



billing_manager_01a

**Light Theme - class btn-default - options disabled**

billing_manager_02

**Light Theme - class btn-default - options enabled**

billing_manager_02a


**Powder Blue Theme - class btn-link - options disabled**

billing_manager_03


**Powder Blue Theme - class btn-link - options enabled**

billing_manager_03a

**Powder Blue Theme - class btn-default - options disabled**

billing_manager_04

**Powder Blue Theme - class btn-default - options enabled**

billing_manager_04a

**Comment by sjpadgett** [comment]
Technically the btn-link class is meant to allow a button to look like a link but still allow the function a button provides. Most developers working with bootstrap would expect this.
The purpose of using bootstrap in the first place is to have a consistent base style and class actions to create themes with. What we have fallen into is overriding bootstrap by redefining styles like buttons btn and btn actions like btn-link focus with almost completely new classes to support these color themes. What should happen is only colors and sizing should change, mostly. Otherwise if we are going to redefine bootstrap classes, why bother with bootstrap at all. It is entirely possible to do responsive in straight html5. Also if we ever decide to do bootstrap 4, this makes upgrading, that much harder.

Light theme is our default production theme and all theming should follow it. Personally, I think color themes make applications look unprofessional and so 1990's but Brady and Doctors like having these different colors to break things up so, who can argue with that! A light or dark theme is all that is needed imo.

In the end, I don't have the energy to argue this or care enough. Follow my advice or don't but it's getting to the point that I now rethink doing any new features because, it's just getting too hard. If I feel this way, think how new developers will feel when they can't use a simple btn-link class.

**Comment by bradymiller** [comment]
hi @sjpadgett and @zbig01 , Very nice to have all these options, but I don't think any options above are wrong or right. @sjpadgett , just pick with your gut with whatever you think looks best in the light theme and move on :)
(@zbig01 , not saying to ignore color themes; it's just that the btn-link in colors will remain consistent since it does the same thing to all btn-link throughout the theme)

**Comment by sjpadgett** [comment]
@stephenwaite @juggernautsei decided to go with this explanation in confirm dialog. Let me know if you have issues or better descriptions. Need to know soon so can bring in for 502
image

**Comment by sjpadgett** [comment]
@bradymiller Sorry, I don't move on. If I think something is going in the wrong direction, I'm going to state it. It's hard enough keeping up with changing api's or learning openemr for new developers to have to also relearn bootstrap.
I can appreciate updating styling but not to the point of redefining basic core bootstrap classes.
Last i'll say on this subject. Now i'll move on:)

**Comment by stephenwaite** [comment]
hi @sjpadgett , agree with you on the clear. You have it working better than before plus new spiffy modal and shed insight onto the inner working of updateclaim billing and claims tables too, great job
:100:

**Comment by sjpadgett** [comment]
Thanks @stephenwaite Really need to dig into postings soon. I'm just not totally convinced we necessarily have many bugs but user misconceptions and need to add some gotcha validations there. 
bringing in. bye

**Pull request merged**